internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+			jobId := armadaevents.UuidFromProtoUuid(e.JobRunSucceeded.GetJobId())
+			batch.JobRunsSucceeded[jobId] = true
+		case *armadaevents.EventSequence_Event_JobRunErrors:
+			eventId := eventid.New(msg.ID(), i).String()
+			for j, jobRunError := range e.JobRunErrors.GetErrors() {
+				bytes, err := proto.Marshal(jobRunError)
+				if err != nil {
+					err = errors.WithStack(err)
+					logging.WithStacktrace(log, err).Error("failed to marshal JobRunError")
+				}
+				batch.JobRunErrors = append(batch.JobRunErrors, JobRunError{
+					// To ensure inserts are idempotent,
+					// we need to mark each row with a deterministic id.
+					ID:       fmt.Sprintf("%s-%d", eventId, j),
+					RunID:    armadaevents.UuidFromProtoUuid(e.JobRunErrors.GetRunId()),
+					Error:    bytes,
+					Terminal: jobRunError.GetTerminal(),
+				})
+			}
+		}
+	}
+	return batches, nil
+}
+
+func (srv *Ingester) ProcessBatches(ctx context.Context, batchChannel <-chan *Batch) error {
+	log := ctxlogrus.Extract(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case batch := <-batchChannel:
+
+			err := srv.ProcessBatch(ctx, batch)
+			if err != nil {
+				return err // TODO: Keep retrying on transient errors.
+			}
+			log.Infof("wrote batch records into postgres")
+
+			// Ack all messages that were used to create the batch.
+			// Acking is only safe once data has been written to postgres.
+			for _, messageId := range batch.MessageIds {
+				err := srv.consumer.AckID(messageId)
+				if err != nil {
+					return errors.WithStack(err)
+				}
+			}
+		}
+	}
+}
+
+func (srv *Ingester) ProcessBatch(ctx context.Context, batch *Batch) error {
+	log := ctxlogrus.Extract(ctx)
+	queries := New(srv.Db)
+
+	// Jobs
+	records := make([]interface{}, len(batch.Jobs))
+	for i, job := range batch.Jobs {
+		records[i] = job
+	}
+	err := Upsert(ctx, srv.Db, "jobs", JobsSchema(), records)
+	if err != nil {
+		return err // TODO: Keep retrying on transient failures.
+	}
+	log.Infof("wrote %d jobs into postgres", len(records))
+
+	// Job runs
+	records = make([]interface{}, len(batch.Runs))
+	for i, run := range batch.Runs {
+		records[i] = run
+	}
+	err = Upsert(ctx, srv.Db, "runs", RunsSchema(), records)
+	if err != nil {
+		return err // TODO: Keep retrying on transient failures.
+	}
+	log.Infof("wrote %d jobs into postgres", len(records))
+
+	// Reprioritisations
+	for _, reprioritisation := range batch.Reprioritisations {
+		for jobSet, priority := range reprioritisation.PrioritiesByJobSet {
+			err := queries.UpdateJobPriorityByJobSet(ctx, UpdateJobPriorityByJobSetParams{
+				JobSet:   jobSet,
+				Priority: priority,
+			})
+			if err != nil {
+				return errors.WithStack(err)
+			}
+		}
+
+		// TODO: This will be slow if there's a large number of ids.
+		// Could be addressed by using a separate table for priority + upsert.
+		for jobId, priority := range reprioritisation.PrioritiesByJob {
+			err := queries.UpdateJobPriorityById(ctx, UpdateJobPriorityByIdParams{
```

Comment:
I think a simple thing here is just to update all priorities in a single statement rather than doing them one by one by one  I'd think we'd need to fix this pretty soon because this will be extremely slow and current workflows do cause large number of reprios to get generated